### ISSUE TITLE
Add promise support

### ISSUE THREAD

**Issue by emgeee** [mixpanel/mixpanel-node#67]
It'd be nice to have the mixpanel methods return promises

**Comment by freeall** [comment]
The good old promises vs callbacks :) I for one like callbacks a lot more

**Comment by tdumitrescu** [comment]
As a "nice to have" API which not everyone wants to use, we'd want to do this in a way that doesn't break backwards compatibility. Perhaps an init option to wrap every method optionally, or a separate object which exposes promise-returning methods. The fastest path to getting it in the lib would be submitting a pull request. In the meanwhile, we'll leave this issue open for now in case others want to chime in.

**Comment by emgeee** [comment]
Any of those solutions work, I was just thinking of having the functions always return a Promise that get's resolve/rejected but then checking for existence of a callback and invoking it if it exists. Something like:

```javascript
function track (cb) {
  return new Promise(function (resolve, reject) {
    trackEvent(function (err, res) {
      if (err) {
         reject(err)
      } else {
        resolve(res)
      }

      if (cb && typeof cb === 'function') cb(err, res)
    }
  }
}
```

Wanted to get some thoughts on implementation before I submit a PR

**Comment by tdumitrescu** [comment]
Sure. You could probably do it by wrapping only `metrics.send_request()` since that's what's used internally for all the async network calls. then the implementation of `track()` etc wouldn't have to change except to end with `return metrics.send_request(endpoint, data, callback);` instead of just `metrics.send_request(endpoint, data, callback);`.

Is there a particular promise lib you'd prefer to use? I don't think we're quite ready yet to require a runtime that supports ES2015 Promises in order to use `mixpanel-node`.

**Comment by emgeee** [comment]
My preference is Bluebird but I've also seen some libraries do a check for native Promise support and then falling back to a pollyfill.

**Comment by tdumitrescu** [comment]
Cool. Have you looked at https://www.npmjs.com/package/native-or-bluebird?

Also it looks like Bluebird offers a `nodeify` method so you don't have to add the callback-checking logic yourself: https://github.com/petkaantonov/bluebird/blob/master/API.md#nodeifyfunction-callback--object-options---promise

If this works without breaking any of the existing unit tests (i.e. the callback API continues to work unchanged), then I don't see any reason not to go for it.

**Comment by tejasmanohar** [comment]
For the time being, you may be able to use Bluebird (or another module's) promisify/promisifyAll to make node-style callbacks -> promise functions

**Comment by tejasmanohar** [comment]
Nodeify is also available as a separate module if BBird is too big for you -> https://www.npmjs.com/package/nodeify

**Comment by duanefields** [comment]
Using bluebird

````coffeescript
token = "whatever"
key = "whatever"
mixpanel_importer = Promise.promisifyAll Mixpanel.init token, { key }
````

**Comment by duanefields** [comment]
See my fix at https://github.com/mixpanel/mixpanel-node/pull/80, or use my fork

**Comment by duanefields** [comment]
Merged my branch with latest changes and fixed unit tests

**Comment by dhritzkiv** [comment]
Any chance this discussion could be revived? Node v7.6.0 supports async-await without runtime flags, and it'd be nice to wait on `mixpanel` methods with await (without using v8.0.0's `util.promisify`)

**Comment by tdumitrescu** [comment]
Indeed, this is easier than ever since all supported versions of Node now have native Promises. Will be revisiting soon. (That being said, most of the time you _don't_ want to suspend execution waiting for the response from a tracking call. I'd imagine that `await mixpanel.track(...)` is rarely useful in practice.)

**Comment by dhritzkiv** [comment]
Cool! And you're right. For most tracked events, I run it without waiting for the result. However, for some tracking, such as for updating people's metadata and tracking revenue, I like to wait on the response.

**Comment by GriffinSauce** [comment]
`await mixpanel.track(...)` is very useful in a script as I've just found out. With async await this change becomes even more valuable because async await eases JS development so much.

**Comment by himanshuphirke** [comment]
@GriffinSauce Does the latest library support `await mixpanel.people.set(...)` ?

**Comment by GriffinSauce** [comment]
It's a while ago but I believe I used bluebirds promisify to get there.

**Comment by himanshuphirke** [comment]
> It's a while ago but I believe I used bluebirds promisify to get there.

thanks

**Comment by henrylearn2rock** [comment]
I tried this: 

```
track = async (event, properties) =>
    new Promise((resolve, eject) =>
        mixpanel.track(event, properties, err => {
            if (err) {
                console.error(err)
                eject(err)
            } else {
                console.log('mixpanel event ok')
                resolve()
            }
        })
    )
```

This does not work when things went ok (no console log) as the `callback` is not even invoked according to https://github.com/mixpanel/mixpanel-node/blob/master/lib/mixpanel-node.js#L139 when things went ok.  Please fix otherwise cannot use this in serverless env because it requires all promises to be resolved in a function.

**Comment by freeall** [comment]
@henrylearn2rock From what you linked to, it seems that `callback(e)` is always invoked.

I noticed that you your method is `async`, but also returns a `new Promise`. Were you maybe meant to only return the promise?

**Comment by vigzmv** [comment]
I am using `mixpanel.people.set` in the backend to update thousands of users, Should I just do it all in go and mixpanel will automatically batch update them in series or do I need to batch and wait for some time gaps?

**Comment by ndrozd** [comment]
Why is this closed? :( 
Bluebird isn't a thing for me and callback seem to be firing without returning any data.

**Comment by izakfilmalter** [comment]
It's 2022 and I need to wrap every method in this lib with new Promise. Your app is only as good as your api that developers have to integrate with. This is why Posthog is killing it right now.

**Comment by steviec** [comment]
It's pretty surprising that promises aren't first class citizens in your library. I'm having to promisify everything to pull it off. Please reconsider your approach to this.

**Comment by dylan-westbury** [comment]
New potential user to Mixpanel, +1 for promise support

**Comment by RohovDmytro** [comment]
+1 from 2025.
